types/react/v16.7/index.d.ts

```diff
@@ -0,0 +1,2519 @@
+         * Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
+         * prevents this from being invoked.
+         *
+         * @deprecated 16.3, use getSnapshotBeforeUpdate instead; will stop working in React 17
+         * @see https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#reading-dom-properties-before-an-update
+         * @see https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path
+         */
+        componentWillUpdate?(nextProps: Readonly<P>, nextState: Readonly<S>, nextContext: any): void;
+        /**
+         * Called immediately before rendering when new props or state is received. Not called for the initial render.
+         *
+         * Note: You cannot call `Component#setState` here.
+         *
+         * This method will not stop working in React 17.
+         *
+         * Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
+         * prevents this from being invoked.
+         *
+         * @deprecated 16.3, use getSnapshotBeforeUpdate instead
+         * @see https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#reading-dom-properties-before-an-update
+         * @see https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path
+         */
+        UNSAFE_componentWillUpdate?(nextProps: Readonly<P>, nextState: Readonly<S>, nextContext: any): void;
+    }
+
+    interface Mixin<P, S> extends ComponentLifecycle<P, S> {
+        mixins?: Array<Mixin<P, S>>;
+        statics?: {
+            [key: string]: any;
+        };
+
+        displayName?: string;
+        propTypes?: ValidationMap<any>;
+        contextTypes?: ValidationMap<any>;
+        childContextTypes?: ValidationMap<any>;
+
+        getDefaultProps?(): P;
+        getInitialState?(): S;
+    }
+
+    interface ComponentSpec<P, S> extends Mixin<P, S> {
+        render(): ReactNode;
+
+        [propertyName: string]: any;
+    }
+
+    function createRef<T>(): RefObject<T>;
+
+    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ComponentType<P & ClassAttributes<T>>;
+
+    //
+    // Hooks
+    // ----------------------------------------------------------------------
+
+    function useState<T>(initialState: T): [T, (newState: T) => void];
+    function useEffect(create: () => (void | (() => void)), inputs?: ReadonlyArray<unknown>): void;
+    function useContext<T>(foo: React.Context<T>): T;
+    function useReducer<S, A>(reducer: (state: S, action: A) => S, initialState: S): [S, (action: A) => void];
+    function useCallback<F extends (...args: never[]) => unknown>(callback: F, inputs?: ReadonlyArray<unknown>): F;
+    function useMemo<T>(create: () => T, inputs?: ReadonlyArray<unknown>): T;
+    function useRef<T extends unknown>(initialValue?: T): React.RefObject<T>;
+    function useImperativeMethods<T>(ref: React.Ref<T>, createInstance: () => T, inputs?: ReadonlyArray<unknown>): void;
```

Comment:
`ref` can be `undefined` (the `React.forwardRef` component can always be just not passed a `ref`; that just means it can forward the `ref` it is given one)

Reply:
```suggestion
    // NOTE: this does not accept strings, but that will have to be fixed by
    // removing strings from type Ref<T> because of how forwardRef is defined.
    /**
     * `useImperativeMethods` customizes the instance value that is exposed to parent components when using
     * `ref`. As always, imperative code using refs should be avoided in most cases.
     *
     * `useImperativeMethods` should be used with `React.forwardRef`.
     *
     * @see https://reactjs.org/docs/hooks-reference.html#useimperativemethods
     */
    function useImperativeMethods<T>(ref: Ref<T>, init: () => T, inputs?: IdentityCheckInputList): void;
```